### ISSUE TITLE
Support parallel execution of tests

### ISSUE THREAD

**Issue by mfulton26** [JetBrains/spek#108]
There are various ways to implement. Personally I believe I prefer to be able to execute `group` and `test` blocks myself using my own `ExecutorService`, etc. but that doesn't work today as Spek "pushes" blocks onto a queue/tree and executes everything itself serially.

One of the big things I don't like about many test frameworks is that they try to provide their own DSL for parallel execution with limited configuration options, etc. instead of simply letting the test developer call the framework using their own parallel/concurrent code so that they can control when/where/how/etc.

JUnit 5 supports practically full dynamic trees of containers (groups) and tests. As such, Spek could potentially be changed to execute groups/tests live instead of "discovering" all tests first and then "executing" the tests separately. This would allow test developers to use for loops, if statements, executor services, etc. to execute their groups and tests as they please.

**Comment by artem-zinnatullin** [comment]
You already can use loops and flow control

**Comment by mfulton26** [comment]
@artem-zinnatullin Yes but to a limited extent. e.g. You can not make a test dependent on results of a previous test or use results of a test to data-drive further tests, etc. As it is today (until Kotlin 1.1 comes out) you cannot do expensive set-up lazily either in a group. Discovering and executing groups and tests at the same time could be very advantageous to give developers further control.

**Comment by artem-zinnatullin** [comment]
Though I find making tests dependent on other tests, dynamic tree will great

**Comment by raniejade** [comment]
AFAIK JUnit 5's dynamic tests doesn't support groups. `ExecutorService`s will just complicate things, since tests are created dynamically we need a reliable way to tell Spek that all tests are generated. You can already use any control flow and with coroutine support in Kotlin `1.1`  we can do a lot more.

**Comment by artem-zinnatullin** [comment]
Oh no, wait, please do not tight Spek with Kotlin's coroutines

On Sat, Aug 27, 2016 at 2:42 PM, Ranie Jade Ramiso <notifications@github.com

**Comment by raniejade** [comment]
@artem-zinnatullin Why is that?

**Comment by artem-zinnatullin** [comment]
Because as @mfulton26 pointed:

>One of the big things I don't like about many test frameworks is that they try to provide their own DSL for parallel execution with limited configuration options, etc. instead of simply letting the test developer call the framework using their own parallel/concurrent code so that they can control when/where/how/etc.

It's better to not tight Spek to specific API for parallelization, for example I (we) would like to use RxJava for that, someone would like to use Executors API, others will use Reactor, etc, etc.

**Comment by raniejade** [comment]
Ahh, yep definitely. What I meant was you can do more complex test setup with coroutines compared to the control flows (which we currently have).

**Comment by artem-zinnatullin** [comment]
Ah, ok!

**Comment by raniejade** [comment]
One thing I can think via extensions that allow to tap into Spek's lifecycle.  Been playing around with extending Spek, here's what I have so far https://github.com/JetBrains/spek/blob/master/spek-junit-platform-engine/src/test/kotlin/org/jetbrains/spek/engine/ExtensionTest.kt.

**Comment by nizienko** [comment]
Hello, I tried to execute tests in parallel with 'maxParallelForks 2' in gradle config, which is works well on simple JUnit5 project without spek. All tests were executed twice in 2 threads.

**Comment by corlaez** [comment]
This feature would be a nice addition!
A button that allows parallel execution would be enough for my needs. Similar to this junit4 plugin: https://plugins.jetbrains.com/plugin/12959-junit4-parallel-runner

**Comment by corlaez** [comment]
https://kotlin.libhunt.com/spek-latest-version

Since 2.0.13 we have these gradle properties that allow parallel execution:

>spek2.discovery.parallel.enabled - controls whether discovery is done in parallel or not, by default it is disabled. The presence of this property regardless of value will enable parallel discovery.

>spek2.execution.parallel.enabled - controls whether execution is done in parallel or not, by default it is disabled. The presence of this property regardless of value will enable parallel execution.

> A note for parallel execution
✅ Parallelism is at a class level not individual test scopes. Spek uses coroutines under the hood to execute tests in parallel, essentially spek2.execution.parallel.enabled just controls which CoroutineDispatcher is used. For parallel executions Dispatchers.Default is used, otherwise the one provided by runBlocking is used.

**Would that be enough to close this issue?**

**Comment by raniejade** [comment]
@corlaez When this ticket was created the plan is to have test scope level parallelism, not just class level. I might have to re-think this a bit further - the real question is: is class/spek level parallelism enough?

**Comment by corlaez** [comment]
got, got it.
Yeah I have noticed that 2.0.13 will execute tests within a class in parallel but at a class level the execution is still sequential.

**Comment by corlaez** [comment]
What I am used to is the way that Spock running with JUni4 Parallel Tests (https://plugins.jetbrains.com/plugin/12959-junit4-parallel-runner) works:
Spock would instantiate the fields every time it is about to run a test. That means that in parallel each test works with its own instances. It sounds like wasted objects and garbage collection triggering more often, and it probably does but it has no impact as we have way more RAM and GC times are negligible. The nice thing is that I don't have to clear state or worry about racing conditions.

I am not even sure that the class parallelism we have right now works like that (or if spek in general is implemented in such a way), I have to do some digging.

**Comment by LisPI** [comment]
> Hello, I tried to execute tests in parallel with 'maxParallelForks 2' in gradle config, which is works well on simple JUnit5 project without spek. All tests were executed twice in 2 threads.

We're experiencing the same issue. Maybe there are any workaround? btw looks like 2.0.4 the first Spek version with the problem
